src/index.js

```diff
@@ -45,21 +45,17 @@ Utils.isDayjs = isDayjs
 Utils.wrapper = wrapper
 
 const parseDate = (date) => {
-  let reg
-  if (date === null) return new Date(NaN) // Treat null as an invalid date
-  if (Utils.isUndefined(date)) return new Date()
+  if (date === null) return new Date(NaN) // null is invalid
+  if (Utils.isUndefined(date)) return new Date() // today
   if (date instanceof Date) return date
-  // eslint-disable-next-line no-cond-assign
-  if ((typeof date === 'string')
-    && (/.*[^Z]$/i.test(date)) // looking for a better way
-    && (reg = date.match(C.REGEX_PARSE))) {
-    // 2018-08-08 or 20180808
-    return new Date(
-      reg[1], reg[2] - 1, reg[3] || 1,
-      reg[4] || 0, reg[5] || 0, reg[6] || 0, reg[7] || 0
-    )
+  if (typeof date === 'string' && !date.toLowerCase().endsWith('z')) {
```

Comment:
Just noticed that `endsWith` is a new added method in js, with cost some more bytes than `/.*[^Z]$/i.test` in the final es5 code.

Reply:
Yes, but the concept is clearer. I also saved some letters on renaming `reg[...]` to `d[...]` multiple times and removing the `let reg`, wouldn't that even it out?

Reply:
`reg` will be minified in building I think, and `endsWith` becomes the added bytes.

Reply:
Funnily, this was never needed at all, there is a bug in the regex.

If you look at it

export const REGEX_PARSE = /^(\d{4})-?(\d{1,2})-?(\d{0,2})[^0-9]*(\d{1,2})?:?(\d{1,2})?:?(\d{1,2})?.?(\d{1,3})?$/i`

you wonder: Why does it even match a `z` or `Z` it in the end? It is because the last dot in the regex is not escaped (the dot in part `?.?(\d{1,3})?`), so that matches *any character* and therefore matches the `z`

If you change it to

export const REGEX_PARSE = /^(\d{4})-?(\d{1,2})-?(\d{0,2})[^0-9]*(\d{1,2})?:?(\d{1,2})?:?(\d{1,2})?\.?(\d{1,3})?$/i`

(note the backlash to escape the dot), all tests are still green if we just do

```javascript
  if (typeof date === 'string') {
    const d = date.match(C.REGEX_PARSE)
    if (d) {
      return new Date(d[1], d[2] - 1, d[3] || 1, d[4] || 0, d[5] || 0, d[6] || 0, d[7] || 0)
    }
  }
```

So we even save a lot of bytes :-).

Reply:
Pushed that fix to PR.

Reply:
Seems we've changed this regex a few days ago, maybe that's the reason?

Reply:
But there are other issues with that check (no matter if we use the new or old regex): For example this date is not parsed by dayjs:

`2018-05-02T11Z` (missing minutes and seconds), but it is parsed by moment.js. Maybe we need a lot more test cases here...

Reply:
Check my latest push - it will fix all of the issues I just mentioned automatically. All we want to do in our custom check is catch the yyyymmdd shorthand so you can parse `20180501`, the rest can be delegated to the `Date()` constructor just like moment.js does. All tests still green.

Reply:
Looks like magic, but could this just because we are lacking of test suits?

Reply:
Yes, but some additional tests I wrote would fail in both cases. Pure Javascript allows for strings without a time separator `T` (so just having a space) or single-digit time or date bits (e.g. `2018-4-1`), there is no need to parse that explicitly - `new Date()` will handle it for us.

But still there is an incosistency with moment.js, as they allow partial time strings, e.g. `2018-01-01 13Z`, which is not ok with dayjs. I could add some (failing) tests to make it clear?

Reply:
No you can't simplify the regex like this. The reason why we are using such long regex is the Date object varies among different browsers, expecially mobile browsers. Then you will facing a lot compatibility problem. 

@iamkun is correct, the magic is just because we are lacking some multi browser tests in a PR checking ( but will do in every CI release)

Reply:
@xxyuk Ok, thanks for the info, I didn't know that this also covers cross-browser concerns. But maybe it would be good to capture the different browser behaviors in unit tests so the developers get a more instant feedback on doing changes?

Reply:
And apart from that: The unescaped dot is still an issue, right?

Reply:
Cross-browser tests cost too much time and resources, that why we only do this on every release.

> The unescaped dot is still an issue, right?

I'm not sure if it's a issue, and my suggestion is you could first revert to the original `regex(Z)`, and let this PR get merged. And open a new PR with this change, we are happy to make a manual multi browser test to see if it's ok

Reply:
I will do that, @xxyuk , good ṕlan.

Reply:
Ok, so I force pushed back to just the initial refactoring. Can we merge that as is? Next up I create a PR where I cut down the regex and escape the dot properly and we run through the CI?

Reply:
Although we are looking for a better way, I think `/.*[^Z]$/i.test(date)` is better here than `endsWith`

The rest looks fine.

Reply:
Ok I did, and even shortened it as `/z$/i` does the same (asserting it ends with a `z`, case insensitive). `.*[^Z]$` was doing too much, no need to match any character 0-n times.

Reply:
`/z$/i.test('z')` doesn't make sense.

Reply:
Yes, I fixed it. I could just use the web view as I do not have the solution checked out right now, will crush those commits, sorry.

Reply:
Needs to be `/[^Z]$`.test(date)`, getting rid of `.*`.

Reply:
Ok, I crushed those commits and no put the regex back. I tried to keep the Regex as short as possible, so I no went for

`!/Z$/.test(date)` rather then `/[^Z]$`i.test(date)`, as it is
- less bytes (not many)
- easier to comprehend the negation outside of the regex instead of with the [^Z] negation syntax. Is that ok now?